fix(compiler): properly type `forwardRef` and `resolveForwardRef`

NickClark opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
#11638

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

Makes `forwardRef` and `resolveForwardRef` return more meaningful  types. Now when resolving a forward reference, TypeScript knows what type the returned reference is.

Closes #11638

vicb commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -32,10 +43,11 @@ export interface ForwardRefFn { (): any; }
-  (<any>forwardRefFn).toString = function() { return stringify(this()); };
-  return (<Type<any>><any>forwardRefFn);
+export function forwardRef<T>(forwardRefFnSrc: ForwardRefFnSrc<T>): ForwardRefFn<T> {
+  let forwardRefFn = <ForwardRefFn<T>>forwardRefFnSrc;
```
const

vicb commented on modules/@angular/forms/src/directives/select_multiple_control_value_accessor.ts:
```diff
@@ -15,7 +15,8 @@ import {ControlValueAccessor, NG_VALUE_ACCESSOR} from './control_value_accessor'
 export const SELECT_MULTIPLE_VALUE_ACCESSOR = {
   provide: NG_VALUE_ACCESSOR,
-  useExisting: forwardRef(() => SelectMultipleControlValueAccessor),
+  useExisting: <ForwardRefFn<typeof SelectMultipleControlValueAccessor>>forwardRef(
```
q: Is the cast really necessary here ?

NickClark replied:
I thought it was. I was getting the error: `SELECT_MULTIPLE_VALUE_ACCESSOR' has or is using name 'ForwardRefFn' from external module "/Users/nick/Code/k3/angular/modules/@angular/core/src/di/forward_ref" but cannot be named.`. But just now, going back I realize now all I needed was to import the forwardRefFn. 😶

vicb commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -18,7 +18,18 @@ import {Type} from '../type';
  * @experimental
  */
-export interface ForwardRefFn { (): any; }
+export interface ForwardRefFnSrc<T> { (): T; }
```
Q: I don't see how this type relates to `ForwardRef`, what about renaming this to `Factory<T>` ?

WDYT

vicb replied:
Do we need this type at all ?

NickClark replied:
No. I just get carried away with naming my types. I was just trying to hint at it's purposed, since it's really just a function that we decorate, and wanted to clearly differentiate it from the type that is returned. 

Shall I just change the `forwardRef` to this: `function forwardRef<T>(forwardRefFnSrc: () => T): ForwardRefFn<T>`? That definitely seems clear enough 😄

vicb replied:
👍

vicb commented:
One minor question, otherwise LGTM

NickClark commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -50,11 +52,14 @@ export function forwardRef(forwardRefFn: ForwardRefFn): Type<any> {
 }
+
+function _isForwardRefFn<T>(t: any): t is ForwardRefFn<T> {
+  return isFunction(t) && t.hasOwnProperty('__forward_ref__') && t.__forward_ref__ === forwardRef;
```
It does seem redundant. I actually just kept the original check here, so I'm not sure if the original author had some reason for it. Is there some reason we need to ensure that the property is not an inherited one?

NickClark replied:
I guess if there's no test for it...

vicb replied:
maybe leave it like this for now... thanks

NickClark commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -50,11 +52,14 @@ export function forwardRef(forwardRefFn: ForwardRefFn): Type<any> {
-      type.__forward_ref__ === forwardRef) {
-    return (<ForwardRefFn>type)();
+export function resolveForwardRef<T>(type: T | ForwardRefFn<T>): T {
+  if (_isForwardRefFn(type)) {
```
ternary: ✅

NickClark commented:
Should be good to go. :shipit:

vicb commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -32,10 +33,11 @@ export interface ForwardRefFn { (): any; }
-  (<any>forwardRefFn).__forward_ref__ = forwardRef;
-  (<any>forwardRefFn).toString = function() { return stringify(this()); };
-  return (<Type<any>><any>forwardRefFn);
+export function forwardRef<T>(forwardRefFnSrc: () => T): ForwardRefFn<T> {
```
Would this work ?

```
export function forwardRef<T>(typeFactory: () => T): ForwardRefFn<T> {
 typeFactory['__forward_ref__'] = forwardRef;
 typeFactory['toString'] = function() { return stringify(this()); };
 return typeFactory;
}
```

vicb replied:
make `__forward_ref__` a const

NickClark replied:
I'll look at this when I get back to my dev computer. Do we allow implicit `any`?

NickClark replied:
Also, if we use the latest TypeScript I think it's supposed to allow something like this now: `forwardRefFn = <ForwardRefFn<T>>forwardRefFn`

vicb replied:
I don't think `['...']` is actually a good idea for minification.

`forwardRefFnSrc` -> `typeFactory` would be nice though.

NickClark replied:
I'm changing the name to `typeFactory`. Not sure I'm clear on what else you're asking for here. Are we accessing object attributes via strings to help prevent minification? Seems like that wouldn't happen.

vicb commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -50,11 +52,10 @@ export function forwardRef(forwardRefFn: ForwardRefFn): Type<any> {
-  } else {
-    return type;
-  }
+export function resolveForwardRef<T>(type: T | ForwardRefFn<T>): T {
```
is this formatted ?

vicb replied:
oh forget about it... that's ok with this clang-format version

NickClark replied:
Ya, the tooling didn't catch it, so I assumed we were good here.  😄

vicb commented:
one minor nit

vicb commented on modules/@angular/core/src/di/forward_ref.ts:
```diff
@@ -50,11 +52,10 @@ export function forwardRef(forwardRefFn: ForwardRefFn): Type<any> {
+}
+
+function _isForwardRefFn<T>(t: any): t is ForwardRefFn<T> {
+  return isFunction(t) && t.hasOwnProperty('__forward_ref__') && t.__forward_ref__ === forwardRef;
```
/cc @mhevery how well do this play with minification ?

vicb commented:
could you please rebase and integrate required changes ?

NickClark commented:
@vicb Ya, sorry about the delay, was gone on vacation.

vicb commented:
No pb Nick, only have time to get back to this PR now

NickClark commented:
Had some questions in regard to the requested refactor of `forwardRef`

vicb commented:
Yep, what ?

NickClark commented:
@vicb :point_up: https://github.com/angular/angular/pull/11813#discussion_r82711327

mhevery commented:
@vicb ping

NickClark commented:
@vicb Hopefully that'll do 😃 :shipit:

tbosch commented:
Hi,
as this is a breaking change we are willing to take in the 5.0 release. Could you rebase against master if you are still interested in landing this PR?

NickClark commented:
Ya, I'll try to get to this in the next day or so.

Commit:
fix(compiler): properly type `forwardRef` and `resolveForwardRef`

Makes `forwardRef` and `resolveForwardRef` return more meaningful  types. Now when resolving a forward reference, TypeScript knows what type the returned reference is.

Closes #11638

NickClark commented:
I believe it's good to go.

NickClark commented:
@tbosch Is anything else needed?

tbosch commented:
We need to create a branch for v5 before we can land this...

splincode commented:
@NickClark @mhevery What is the status of the task?

alexeagle commented:
Closing pull requests from 2016, sorry this got stale and was never merged. Please open a new one if you still care about this change.
